shapely/coordinates.py

```diff
@@ -54,20 +148,117 @@ def transform(geometry, transformation, include_z=False):
     if new_coordinates.dtype != np.float64:
         raise ValueError(
             "The provided transformation returned an array with an unexpected "
-            f"dtype ({new_coordinates.dtype})"
+            f"dtype ({new_coordinates.dtype}, but expected {coordinates.dtype})"
         )
     if new_coordinates.shape != coordinates.shape:
         # if the shape is too small we will get a segfault
         raise ValueError(
             "The provided transformation returned an array with an unexpected "
-            f"shape ({new_coordinates.shape})"
+            f"shape ({new_coordinates.shape}, but expected {coordinates.shape})"
         )
     geometry_arr = lib.set_coordinates(geometry_arr, new_coordinates)
     if geometry_arr.ndim == 0 and not isinstance(geometry, np.ndarray):
         return geometry_arr.item()
     return geometry_arr
 
 
+def _transform_multi_vec(geometry, transformation, include_z: bool = False):
+    try:
+        # First we try to apply func to x, y, z vectors.
+        return _transform(
+            geometry,
+            lambda coords: np.array(transformation(*coords.T)).T,
+            include_z=include_z,
+        )
+    except Exception:
+        # A func that assumes x, y, z are single values will likely raise a
+        # TypeError or a ValueError in which case we'll try again.
```

Comment:
I am personally not a huge fan of such an automatic fallback. I know this exists in `shapely.ops.transform`. But if we would start from scratch (what we somewhat did for `shapely.transform`, and if we don't try to back `shapely.ops.transform` by `shapely.transform` with different defaults), I am not sure I would add this. If a user needs this, they can still write a small wrapper around the scalar function. 

(but I don't have a good sense for what `shapely.ops.transform` currently gets used, and with what kind of functions)

Reply:
> I am personally not a huge fan of such an automatic fallback. I know this exists in `shapely.ops.transform`. But if we would start from scratch (what we somewhat did for `shapely.transform`, and if we don't try to back `shapely.ops.transform` by `shapely.transform` with different defaults), I am not sure I would add this. If a user needs this, they can still write a small wrapper around the scalar function.
> 
> (but I don't have a good sense for what `shapely.ops.transform` currently gets used, and with what kind of functions)

Well, I wouldn't use this fallback either, but I wanted to maintain full backwards compatibility. I also timed it figured that adding the `try` `except` wouldn't be a performance penalty for the first fast option, and if you ended up using the fallback function then I guess you don't care about performance anyway so that's also fine... 

Do we want eventually one function that has all the options with full backwards compatibility ? I think this way people can choose whichever combination of parameters they want as there is one function that supports them all.

If no, then I'm happy with removing this fallback, although I'm not convinced that there is any major drawback of keeping it.

Reply:
I follow the reasoning and I am :+1: on implementing it this way. However, because it is an implementation that is not a "common practice", I think it the reasoning behind it should be annotated.

Also, the same implemtation occurs later in `transform_rebuild_planar`. Maybe you can factor this out in a helper function like

```
def _run_maybe_vectorized_transformation(transform, geometry, transformation, coords, include_z):
    ...
```

This would include the reasoning behind the `try ... except`

Reply:
Additionally, the error would occur in the `transformation` function. So the try ... except (potentially) catches too much now.

Maybe a standard function wrapper?

```
def transpose_transformation(transformation):
    @functools.wraps(func)
    def wrapper(coords):
        try:
            return np.array(transformation(*coords.T).T
        except Exception:
            return np.array([transformation(*c) for c in coords])


# in the transform function do:
transformation = transpose_transformation(transformation) if not interleaved else transformation
```